File: dg-cinema-presentation/src/test/java/devgraft/dgcinema/presentation/adapter/in/query/MovieQueryApiTest.java
package devgraft.dgcinema.presentation.adapter.in.query;

import devgraft.dgcinema.domain.model.Movie;
import devgraft.dgcinema.domain.model.MovieFixture;
import devgraft.dgcinema.domain.port.in.MovieSearchUseCase;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import static org.mockito.ArgumentMatchers.anyLong;
import org.mockito.BDDMockito;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.time.format.DateTimeFormatter;
import java.util.List;

@ExtendWith(MockitoExtension.class)
class MovieQueryApiTest {
    private MockMvc mockMvc;
    @InjectMocks
    private MovieQueryApi movieQueryApi;
    @Mock
    private MovieSearchUseCase mockMovieSearchUseCase;
    @Captor
    private ArgumentCaptor<Long> movieIdCaptor;

    @BeforeEach
    void setUp() {
        mockMvc = MockMvcBuilders.standaloneSetup(movieQueryApi).build();
    }

    @Test
    void searchMovie_status_is_ok() throws Exception {
        BDDMockito.given(mockMovieSearchUseCase.getMovie(anyLong())).willReturn(MovieFixture.anMovie().build());

        mockMvc.perform(MockMvcRequestBuilders.get("/movies/{movieId}", 1))
                .andExpect(MockMvcResultMatchers.status().isOk());
    }

    @Test
    void searchMovie_return_value() throws Exception {
        final Movie givenMovie = MovieFixture.anMovie().build();
        BDDMockito.given(mockMovieSearchUseCase.getMovie(anyLong())).willReturn(givenMovie);

        mockMvc.perform(MockMvcRequestBuilders.get("/movies/{movieId}", 1))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movieId").value(givenMovie.id()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.title").value(givenMovie.title()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.description").value(givenMovie.description()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.bannerUrl").value(givenMovie.bannerUrl()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.genre").value(givenMovie.genre()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.releaseDate").value(givenMovie.releaseDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))))
                .andExpect(MockMvcResultMatchers.jsonPath("$.duration").value(givenMovie.duration()))
        ;
    }

    @Test
    void searchMovie_passes_movieId_to_useCase() throws Exception {
        final Long givenMovieId = 1L;
        BDDMockito.given(mockMovieSearchUseCase.getMovie(anyLong())).willReturn(MovieFixture.anMovie().build());

        mockMvc.perform(MockMvcRequestBuilders.get("/movies/{movieId}", givenMovieId));

        Mockito.verify(mockMovieSearchUseCase, Mockito.times(1)).getMovie(movieIdCaptor.capture());
        Assertions.assertThat(movieIdCaptor.getValue()).isNotNull();
        Assertions.assertThat(movieIdCaptor.getValue()).isEqualTo(givenMovieId);

    }

    @Test
    void searchMovies_status_is_ok() throws Exception {
        BDDMockito.given(mockMovieSearchUseCase.getMovieList()).willReturn(List.of(MovieFixture.anMovie().build()));

        mockMvc.perform(MockMvcRequestBuilders.get("/movies"))
                .andExpect(MockMvcResultMatchers.status().isOk());
    }

    @Test
    void searchMovies_return_value() throws Exception {
        final Movie givenMovie = MovieFixture.anMovie().build();
        BDDMockito.given(mockMovieSearchUseCase.getMovieList()).willReturn(List.of(givenMovie));

        mockMvc.perform(MockMvcRequestBuilders.get("/movies"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies").isArray())
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].movieId").value(givenMovie.id()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].title").value(givenMovie.title()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].description").value(givenMovie.description()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].bannerUrl").value(givenMovie.bannerUrl()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].genre").value(givenMovie.genre()))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].releaseDate").value(givenMovie.releaseDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))))
                .andExpect(MockMvcResultMatchers.jsonPath("$.movies[0].duration").value(givenMovie.duration()))
        ;
    }

    @Test
    void searchMovie_was_call_useCase() throws Exception {
        BDDMockito.given(mockMovieSearchUseCase.getMovieList()).willReturn(List.of(MovieFixture.anMovie().build()));

        mockMvc.perform(MockMvcRequestBuilders.get("/movies"));

        Mockito.verify(mockMovieSearchUseCase, Mockito.times(1)).getMovieList();
    }
}

File: dg-cinema-presentation/src/test/kotlin/devgraft/dgcinema/adapter/in/query/MovieQueryApiTest.kt
package devgraft.dgcinema.adapter.`in`.query

import capture
import devgraft.dgcinema.RestDocsApiTest
import devgraft.dgcinema.domain.model.anMovie
import devgraft.dgcinema.domain.ports.`in`.query.MovieSearchUseCase
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.BDDMockito
import org.mockito.Captor
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.anyLong
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.springframework.restdocs.mockmvc.RestDocumentationRequestBuilders.get
import org.springframework.restdocs.payload.JsonFieldType
import org.springframework.restdocs.payload.PayloadDocumentation.fieldWithPath
import org.springframework.restdocs.payload.PayloadDocumentation.responseFields
import org.springframework.restdocs.request.RequestDocumentation.parameterWithName
import org.springframework.restdocs.request.RequestDocumentation.pathParameters
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class MovieQueryApiTest : RestDocsApiTest() {
    @InjectMocks
    private lateinit var movieQueryApi: MovieQueryApi

    @Mock
    private lateinit var mockMovieSearchUseCase: MovieSearchUseCase

    @Captor
    private lateinit var movieIdCaptor: ArgumentCaptor<Long>
    override fun api() = arrayOf<Any>(movieQueryApi)

    @Test
    fun searchMovie_status_is_ok() {
        BDDMockito.given(mockMovieSearchUseCase.getMovie(anyLong())).willReturn(anMovie().build())

        mockMvc.perform(get("/movies/{movieId}", 1))
            .andExpect(status().isOk)
    }

    @Test
    fun searchMovie_return_value() {
        val givenMovie = anMovie().title("givenTitle").description("givenDescription").build()
        BDDMockito.given(mockMovieSearchUseCase.getMovie(anyLong())).willReturn(givenMovie)

        mockMvc.perform(get("/movies/{movieId}", 1))
                .andExpect(status().isOk)
            .andExpect(jsonPath("$.movieId").value(givenMovie.id))
            .andExpect(jsonPath("$.title").value(givenMovie.title))
            .andExpect(jsonPath("$.description").value(givenMovie.description))
            .andExpect(jsonPath("$.bannerUrl").value(givenMovie.bannerUrl))
            .andExpect(jsonPath("$.genre").value(givenMovie.genre))
            .andExpect(jsonPath("$.releaseDate").value(givenMovie.releaseDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))))
            .andExpect(jsonPath("$.duration").value(givenMovie.duration))
            .andDo(
                document(
                    pathParameters(
                        parameterWithName("movieId").description("영화 아이디")
                    ),
                    responseFields(
                        fieldWithPath("movieId").type(JsonFieldType.NUMBER).description("영화 아이디"),
                        fieldWithPath("title").type(JsonFieldType.STRING).description("영화 제목"),
                        fieldWithPath("description").type(JsonFieldType.STRING).description("영화 설명"),
                        fieldWithPath("bannerUrl").type(JsonFieldType.STRING).description("영화 배너 URL"),
                        fieldWithPath("genre").type(JsonFieldType.STRING).description("영화 장르"),
                        fieldWithPath("releaseDate").type(LocalDate::class).description("영화 개봉일"),
                        fieldWithPath("duration").type(JsonFieldType.NUMBER).description("영화 길이(1분)")
                    )
                )
            )
    }

    @Test
    fun searchMovie_passes_parameter_to_useCase() {
        BDDMockito.given(mockMovieSearchUseCase.getMovie(anyLong())).willReturn(anMovie().build())

        mockMvc.perform(get("/movies/{movieId}", 1))

        verify(mockMovieSearchUseCase, times(1)).getMovie(capture(movieIdCaptor))
        Assertions.assertThat(movieIdCaptor.value).isEqualTo(1)
    }

    @Test
    fun searchMovieList_status_is_ok() {
        mockMvc.perform(get("/movies"))
            .andExpect(status().isOk)
    }

    @Test
    fun searchMovieList_return_value() {
        val givenMovie = anMovie().build()
        BDDMockito.given(mockMovieSearchUseCase.getMovieList()).willReturn(listOf(givenMovie))

        mockMvc.perform(get("/movies"))
            .andExpect(jsonPath("$.movies").isArray)
            .andExpect(jsonPath("$.movies[0].movieId").value(givenMovie.id))
            .andExpect(jsonPath("$.movies[0].title").value(givenMovie.title))
            .andExpect(jsonPath("$.movies[0].description").value(givenMovie.description))
            .andExpect(jsonPath("$.movies[0].bannerUrl").value(givenMovie.bannerUrl))
            .andExpect(jsonPath("$.movies[0].genre").value(givenMovie.genre))
            .andExpect(jsonPath("$.movies[0].duration").value(givenMovie.duration))
            .andExpect(
                jsonPath("$.movies[0].releaseDate").value(
                    givenMovie.releaseDate.format(
                        DateTimeFormatter.ofPattern(
                            "yyyy-MM-dd"
                        )
                    )
                )
            )
    }
}
